### PR TITLE
remove GA feature-gate for 1.28

### PR DESCRIPTION
**Pull request by cyclinder**
Signed-off-by: cyclinder <qifeng.guo@daocloud.io>

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

remove GA feature-gate for 1.26

#### Does this PR introduce a user-facing change?

```release-note
The feature gates `DisableAcceleratorUsageMetrics` and `PodSecurity` that graduated to GA and were unconditionally enabled have been removed in v1.28
```

### CODE REVIEW

**Comment by aojea** [comment]
is this not out of the release window and should go in 1.27?

**Comment by cyclinder** [comment]
yeah, it's should be appear in 1.27

**Comment by sftim** [comment]
/remove-label release-note-none

This change is user visible and would need a release note, or an _explanation_ (eg in a comment) of why not.

**Review by jiahuif — CHANGES_REQUESTED** [review]
- I see some feature gates being added back. Could you check if there is any problem with the git rebase?
- I don't agree that there is no "user-facing change". It would be better to mention removal of feature gates in release notes.
- `CSIMigrationAWS` is safe to remove because it was locked to default anyways.

**Comment by bart0sh** [comment]
@cyclinder please, fix failed CI tests

**Comment by cyclinder** [comment]
@bart0sh @jiahuif sorry for the delay. I'll fix these ci tests.

**Comment by pacoxu** [comment]
PR has a merge conflict.

**Review by jiahuif — CHANGES_REQUESTED** [review]

**Review by pacoxu — APPROVED** [review]
/lgtm
/retest

**Comment by bart0sh** [comment]
@cyclinder This is a user visible change. Please add release note.
/hold

File: pkg/features/kube_features.go
Comment (pacoxu) [inline pkg/features/kube_features.go:854]: /cc @ffromani 
Should this be removed in v1.27 or later?

**Comment by pacoxu** [comment]
/release-note-edit
```release-note
The feature gates `CSIMigrationAWS`, `CSIMigrationAzureDisk`, `CSIMigrationGCE`, `CSIStorageCapacity`, `ControllerManagerLeaderMigration`, `DevicePlugins`, `DisableAcceleratorUsageMetrics` and `PodSecurity` that graduated to GA and were unconditionally enabled have been removed in v1.27 
```

DevicePlugins  is GAed in v1.26. Should this be removed in v1.27?
/cc @swatisehgal

File: pkg/features/kube_features.go
Comment (pacoxu) [reply pkg/features/kube_features.go:854]: CPUManager is not removed completely if needed. 
/lgtm cancel

**Comment by cyclinder** [comment]
> @cyclinder This is a user visible change. Please add release note. /hold

I'm sorry I didn't notice this, Now I add release note

File: pkg/features/kube_features.go
Comment (cyclinder) [reply pkg/features/kube_features.go]: Thanks, This is because a this feature-gate was removed last night

File: pkg/features/kube_features.go
Comment (swatisehgal) [inline pkg/features/kube_features.go:117]: Why is this not removed? Same with `CSIMigrationvSphere` below?

File: pkg/features/kube_features.go
Comment (swatisehgal) [inline pkg/features/kube_features.go:161]: Why is this not removed?

File: pkg/features/kube_features.go
Comment (swatisehgal) [inline pkg/features/kube_features.go:178]: Same as above.

**Review by swatisehgal — CHANGES_REQUESTED** [review]
The PR description says that we are removing references to features that graduated to GA in 1.26 but I see some (not all) feature gates were removed. Examples of features that graduated to GA in 1.26 that were not removed are: `CSIMigrationvSphere`, `CSIMigrationAzureFile`, `DelegateFSGroupToCSIDriver`,   `EndpointSliceTerminatingConditionwhich`, `JobTrackingWithFinalizers`,  `KubeletCredentialProviders`, `LegacyServiceAccountTokenNoAutoGeneration`, `MixedProtocolLBService`. Any reason why they were skipped ?

It is the same case with features that graduated to GA in in 1.24 and 1.25. Some left and some removed from `kube_features.go`. I also noticed references to features that graduated to GA in 1.20!!

Can you please update the PR description to clearly describe the scope of this PR. Is it handling just features corresponding to 1.26 or includes the ones that graduated in prior releases and update the PR accordingly?

File: pkg/features/kube_features.go
Comment (ffromani) [reply pkg/features/kube_features.go:854]: I think this should be removed anytime starting 1.27, yes. There are other FG referencing CPUManager, but those are indpeendent and it's fine to leave them there.

File: pkg/features/kube_features.go
Comment (ffromani) [inline pkg/features/kube_features.go:855]: @cyclinder please remove the other bits pertaining to CPUManager above (around lines 70-75)

File: pkg/features/kube_features.go
Comment (cyclinder) [reply pkg/features/kube_features.go:117]: Thanks for your reply!  I noticed it should be removed in 1.28. https://github.com/kubernetes/kubernetes/blob/e18fa0075754aa49d085826a2e0de96cf5fb5144/pkg/features/kube_features.go#L796

File: pkg/features/kube_features.go
Comment (cyclinder) [reply pkg/features/kube_features.go:161]: CI always failed when I removed this feature-gate. I try to figure out why it failed

File: pkg/features/kube_features.go
Comment (cyclinder) [reply pkg/features/kube_features.go:178]: Same as `CSIMigrationAzureFile `, It should be removed in 1.28, So maybe it should wait until 1.28 ?

File: pkg/features/kube_features.go
Comment (cyclinder) [reply pkg/features/kube_features.go:855]: Oh thanks, I miss it

**Review by jiahuif — APPROVED** [review]
/lgtm

File: pkg/features/kube_features.go
Comment (bart0sh) [reply pkg/features/kube_features.go:855]: /unhold

**Review by pacoxu — APPROVED** [review]
/assign @dims @liggitt

**Review by ffromani — COMMENTED** [review]
/hold
unless I miss something, references to the CPUManager FG around line 70 are still there? Let's doublecheck please. Feel free to unhold anytime when this is verified.

File: pkg/features/kube_features.go
Comment (pacoxu) [reply pkg/features/kube_features.go:854]: image
Here should be removed as well.


Sorry, miss that. @cyclinder  would you update it?.

**Comment by dims** [comment]
please poke me after getting LGTM from some of the reviewers from storage, node sigs. thanks!

File: pkg/features/kube_features.go
Comment (pacoxu) [reply pkg/features/kube_features.go:854]: /unhold 
for CPUManager part update

File: pkg/features/kube_features.go
Comment (pacoxu) [inline pkg/features/kube_features.go:125]: /cc andyzhangx leakingtapan pohly

@andyzhangx CSIMigrationAzureDisk
@leakingtapan CSIMigrationAWS
@pohly CSIStorageCapacity

File: pkg/features/kube_features.go
Comment (andyzhangx) [reply pkg/features/kube_features.go:170]: that's correct, `CSIMigrationAzureFile` should be removed in 1.28

File: cmd/kubelet/app/plugins_providers.go
Comment (andyzhangx) [inline cmd/kubelet/app/plugins_providers.go:74]: this means the azure disk in-tree plugin does not register any more.
if there is an azure disk pv in k8s 1.23 cluster for example, then user upgrades to 1.27 directly, that in-tree pv would never work again unless user delete that pv and creates a csi driver pv.

I think even we remove the in-tree driver, those legacy in-tree drivers should still be registered, the driver requests would be forwarded to csi driver to handle.

cc @msau42 @jsafrane

Comment (andyzhangx) [reply cmd/kubelet/app/plugins_providers.go:74]: this issue also exists for gcepd and awsebs, is the above behavior wanted? IMO we should register those intree drivers as a plance holder driver, and csi driver would handle the in-tree requests finally. @jingxu97

Comment (liggitt) [reply cmd/kubelet/app/plugins_providers.go:74]: > if there is an azure disk pv in k8s 1.23 cluster for example, then user upgrades to 1.27 directly...

FWIW, skip-level control-plane upgrades are _explicitly_ unsupported. see https://kubernetes.io/releases/version-skew-policy/#kube-apiserver-1

>> Note: Project policies for [API deprecation](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) and [API change guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md) require kube-apiserver to not skip minor versions when upgrading, even in single-instance clusters.

> ...that in-tree pv would never work again unless user delete that pv and creates a csi driver pv.

what process between 1.23 and 1.27 handles the in-tree → CSI migration?

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi_test.go
Comment (liggitt) [inline pkg/scheduler/framework/plugins/nodevolumelimits/non_csi_test.go:152]: these test changes (without any visible change to feature gate status in this test) are surprising to me

File: pkg/kubelet/cadvisor/cadvisor_linux.go
Comment (liggitt) [inline pkg/kubelet/cadvisor/cadvisor_linux.go:98]: I think this is dropped in https://github.com/kubernetes/kubernetes/pull/114883, right?

File: cmd/kubelet/app/plugins_providers.go
Comment (andyzhangx) [reply cmd/kubelet/app/plugins_providers.go:74]: it would also break if upgrades 1.26 to 1.27 for in-tree pv, as long as it's on 1.27 with this PR, in-tree pv does not work any more since the legacy in-tree driver name is not registered any more.

Comment (jsafrane) [reply cmd/kubelet/app/plugins_providers.go:74]: In a cluster with kube-apiserver and kube-controller-manager 1.27, there can be kubelets with ver. 1.27, 1.26 and 1.25 (see [version skew policy](https://kubernetes.io/releases/version-skew-policy/#kubelet)). CSI migration of AzureDisk, GCE PD and AWS EBS [are GA in 1.25](https://github.com/kubernetes/kubernetes/blob/3534f896e9ea6936b50d821954374374f41c58bd/pkg/features/kube_features.go#L917).

Therefore all nodes in the cluster + KCM + API server have CSI migration GA and in-tree volume plugins are not registered in any component.

AzureFile + vSphere CSI driver feature gates must not be removed, they were GA in 1.26. And they are not removed in this PR.

Comment (jsafrane) [reply cmd/kubelet/app/plugins_providers.go:74]: > it would also break if upgrades from 1.26 to 1.27 for in-tree pv, as long as it's on 1.27 with this PR, in-tree pv does not work any more since the legacy in-tree driver name is not registered any more.

In-tree PVs are quietly translated to CSI and attached/mounted by the CSI driver. This translation is in-memory only, the PVs still look like in-tree in the API server.

**Comment by bart0sh** [comment]
@cyclinder please fix failing CI tests, thanks

File: pkg/scheduler/framework/plugins/nodevolumelimits/non_csi_test.go
Comment (cyclinder) [reply pkg/scheduler/framework/plugins/nodevolumelimits/non_csi_test.go]: 😄 yeah, No test changes should be made here

File: pkg/kubelet/cadvisor/cadvisor_linux.go
Comment (cyclinder) [reply pkg/kubelet/cadvisor/cadvisor_linux.go]: In fact, this PR does not involve the above changes, right?

**Comment by cyclinder** [comment]
> @cyclinder please fix failing CI tests, thanks

thanks for your watching! now i'm focus on these failing CI tests

File: cmd/kubelet/app/plugins_providers.go
Comment (cyclinder) [reply cmd/kubelet/app/plugins_providers.go:74]: thanks, I'm not sure about this part of the code, so we should keep these changes here?

**Comment by cyclinder** [comment]
Thanks @andyzhangx . all tests passed, Is it time to get lgtm?

File: cmd/kubelet/app/plugins_providers.go
Comment (jsafrane) [reply cmd/kubelet/app/plugins_providers.go:74]: I think the PR is OK, at least now.

Comment (cyclinder) [reply cmd/kubelet/app/plugins_providers.go:74]: Thanks @jsafrane

**Review by jiahuif — COMMENTED** [review]
Could you remove `ControllerManagerLeaderMigration` from release notes? This feature gate was already removed by another PR. Or better, could you look at each feature gate in the release notes to ensure that matches the code? I see only `ControllerManagerLeaderMigration` but it is still worthy to double check. Code lgtm.

**Comment by cyclinder** [comment]
Thanks @jiahuif ,  Updated. I remove `ControllerManagerLeaderMigration`  from release notes , and add `CPUManager ` to it(missing)

**Review by jiahuif — APPROVED** [review]
/lgtm

File: pkg/features/kube_features.go
Comment (SergeyKanzhelev) [inline pkg/features/kube_features.go:226]: this says `GA: v1.26`. Typically we wait for two releases to remove the feature gate

Comment (cyclinder) [reply pkg/features/kube_features.go:226]: Thanks for your point @SergeyKanzhelev, Actually, this feature gate doesn't have any code references, so I think it's reasonable to remove it.

Comment (SergeyKanzhelev) [reply pkg/features/kube_features.go:226]: this is by design. Feature Gate stays in the list of feature gates for two releases so environments where it is configured will not break. See https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior

/lgtm cancel

Comment (cyclinder) [reply pkg/features/kube_features.go:226]: Thanks! Now updated.

File: pkg/features/kube_features.go
Comment (SergeyKanzhelev) [inline pkg/features/kube_features.go:75]: keep it since it was GA-d in 1.26. We need to keep it for two versions

File: pkg/features/kube_features.go
Comment (SergeyKanzhelev) [reply pkg/features/kube_features.go:856]: policy asks to keep the feature gate for 2 releases. See https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior

File: pkg/features/kube_features.go
Comment (cyclinder) [reply pkg/features/kube_features.go]: Sorry for reply,  Done.

File: pkg/volume/csimigration/plugin_manager_test.go
Comment (liggitt) [inline pkg/volume/csimigration/plugin_manager_test.go:78]: shouldn't we keep this test case and ensure the expected migration behavior still occurs even without any feature gate fiddling?

File: pkg/volume/csimigration/plugin_manager_test.go
Comment (liggitt) [inline pkg/volume/csimigration/plugin_manager_test.go:201]: changing this to `if len(test.pluginFeature) > 0` would let us blank out `pluginFeature` for testcases for permanent on / locked feature gates and still test the migration behavior

Comment (cyclinder) [reply pkg/volume/csimigration/plugin_manager_test.go:201]: Thanks for the point! Done.

File: pkg/scheduler/framework/plugins/nodevolumelimits/utils.go
Comment (SergeyKanzhelev) [inline pkg/scheduler/framework/plugins/nodevolumelimits/utils.go:43]: is it suppose to be `true` or `false`? Does it have to return at all? Today the feature gate is locked to `true`, meaning we will not enter this if statement and will just go forward executing the rest of the function without returning from here

Comment (cyclinder) [reply pkg/scheduler/framework/plugins/nodevolumelimits/utils.go:43]: Yes, it's supposed to be `true`,  I think it's the same on https://github.com/kubernetes/kubernetes/pull/115838/files

Comment (liggitt) [reply pkg/scheduler/framework/plugins/nodevolumelimits/utils.go]: cc @torredil @dims ^

Comment (liggitt) [reply pkg/scheduler/framework/plugins/nodevolumelimits/utils.go]: looks like https://github.com/kubernetes/kubernetes/pull/115838/files#diff-6901af4ac953694f828dd8bac6f624717b2f622a143f293b9258a46f9a2a559fR43 locked it to true

that seems correct... if asked, AWS migration *is* on

Comment (dims) [reply pkg/scheduler/framework/plugins/nodevolumelimits/utils.go]: Let's ask the folks who know best... @jsafrane @xing-yang What should it be here?

Comment (torredil) [reply pkg/scheduler/framework/plugins/nodevolumelimits/utils.go]: This should be set to `true`.

Comment (SergeyKanzhelev) [reply pkg/scheduler/framework/plugins/nodevolumelimits/utils.go]: is it not working as it should be working on 1.26 where the feature gate is locked to true?

Comment (cyclinder) [reply pkg/scheduler/framework/plugins/nodevolumelimits/utils.go]: I think it is. If there needs a double-checked, Let's ask someone who knows... 

/cc @torredil @jsafrane thanks

Comment (SergeyKanzhelev) [reply pkg/scheduler/framework/plugins/nodevolumelimits/utils.go]: > Yes, this code chunk is correct. `isCSIMigrationOn()` should always return `true` for AzureDisk and GCE PD, because their CSI migration was GA and locked and they're being removed here.

I am asking about 1.26 release. Was it suppose to return true when `utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationAWS)` is true?

Comment (SergeyKanzhelev) [reply pkg/scheduler/framework/plugins/nodevolumelimits/utils.go]: I understand that the unsupported skews are not supported. But would it make sense to keep this logic for longer? What the worst can happen if we are running with older than supported nodes and how easy it will be discovered by the end user?

Comment (msau42) [reply pkg/scheduler/framework/plugins/nodevolumelimits/utils.go]: `CSIMigrationGCE` is GA + lock to true in 1.25: https://github.com/kubernetes/kubernetes/blob/release-1.25/pkg/features/kube_features.go#L923

**Comment by andyzhangx** [comment]
I have triggered azure in-tree e2e tests to make sure there would not be regression for this PR.

**Comment by cyclinder** [comment]
Is this [CI flaky](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/114068/pull-kubernetes-e2e-gce-cos-alpha-features/1635545006783598592/ ) related to the changes?

**Comment by pacoxu** [comment]
> Is this [CI flaky](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/114068/pull-kubernetes-e2e-gce-cos-alpha-features/1635545006783598592/) related to the changes?

This is a failing test https://github.com/kubernetes/kubernetes/issues/116515.

**Comment by humblec** [comment]
> > Is this [CI flaky](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/114068/pull-kubernetes-e2e-gce-cos-alpha-features/1635545006783598592/) related to the changes?
> 
> This is a failing test #116515.

Should be good with https://github.com/kubernetes/kubernetes/pull/116578

**Comment by liggitt** [comment]
@msau42 @jsafrane - is it better to put the CSI migration gate removals in a PR that also cleans up the in-tree volume plugins for those types? AWS gate was dropped in https://github.com/kubernetes/kubernetes/pull/115838 with the accompanying code, and that is a little easier to reason about (the gate went away when the code did)

**Comment by msau42** [comment]
Is it better to put the CSI migration gate removals in a PR that also cleans up the in-tree volume plugins for those types? 

I agree it makes more logical sense to combine them although is not strictly required.

**Comment by msau42** [comment]
> Is it better to put the CSI migration gate removals in a PR that also cleans up the in-tree volume plugins for those types?

 I agree it makes more logical sense to combine them although is not strictly required.

**Comment by cyclinder** [comment]
Thanks @liggitt @msau42 .

I opened a pr (https://github.com/kubernetes/kubernetes/pull/117055) and put the CSI migration gate removals in https://github.com/kubernetes/kubernetes/pull/117055.

This PR only cleans up these gates: `DisableAcceleratorUsageMetrics` and `PodSecurity`.  Would you mind taking a review?

**Comment by liggitt** [comment]
lgtm for PodSecurity for sig-auth

**Comment by pacoxu** [comment]
/release-note-edit
```release-note
The feature gates `DisableAcceleratorUsageMetrics` and `PodSecurity` that graduated to GA and were unconditionally enabled have been removed in v1.28
```

**Pull request merged**
